include/Signal.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+#include <memory>
+#include <vector>
+#include <functional>
+
+namespace Framework
+{
+
+	template<typename> class CSignal;
+	template<typename T, typename... Args>
+	class CSignal<T (Args...)>
+	{
+
+	public:
+		class CConnection;
+
+		typedef std::shared_ptr<CConnection> CConnectionPtr;
+		typedef std::weak_ptr<CConnection> CConnectionWeakPtr;
+		typedef std::vector<CConnectionPtr> CConnectionListPtr;
+
+		typedef CSignal<T (Args...)> CSignalProto;
+		typedef std::function<void()> CConectionFunction;
+		typedef std::function<T (Args...)> CSlotFunction;
+		typedef std::weak_ptr<const CSlotFunction> CSlotFunctionPtr;
+
+		CSignal() = default;
+
+		CConnectionPtr connect(const CSlotFunction& func, bool oneShot = false)
+		{
+			if(func)
+			{
+				std::unique_lock<std::mutex> lock(m_lock);
+
+				CConnectionPtr connection = std::make_shared<CConnection>(func, oneShot);
+				m_connections.push_back(connection);
+
+				return connection;
+			}
+
+			return nullptr;
+		}
+
+		CConnectionPtr connectOnce(const CSlotFunction& func)
+		{
+			return connect(func, true);
+		}
+
+		void Reset()
+		{
+			m_connections.clear();
+		}
+		void operator()(Args... args)
+		{
+			m_connections.erase(
+				std::remove_if(
+					m_connections.begin(), 
+					m_connections.end(),
+					[&](CConnectionWeakPtr& connection)
+					{
+						auto connectionPtr = connection.lock();
+						if(connectionPtr)
+							(*connectionPtr)(args...);
+						return !connectionPtr || (*connectionPtr).isOneShot() || !(*connectionPtr).isConnected();
```

Comment:
Is it necessary to check if the connection is connected here? If it's one shot, the connection will be removed.

Reply:
you're right, and we can assume if the code made it this far, then the `shared_ptr` still exist, else it return true in the 1st condition